File: forum/src/main/java/com/example/forum/error/ExceptionHandlingController.java
package com.example.forum.error;

import com.example.forum.exception.UserAlreadyExistsException;
import com.example.forum.exception.UserNotFoundException;
import com.example.forum.exception.UserPasswordNotEqualsException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.List;
import java.util.stream.Collectors;

@RestControllerAdvice
public class ExceptionHandlingController {
    private final Logger log = LoggerFactory.getLogger(ExceptionHandlingController.class);

    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected ErrorResponse handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
        log.error("input value invalidation");
        List<ErrorResponse.FieldError> fieldErrors = getFieldErrors(e.getBindingResult());
        return buildFieldErrors(ErrorCode.INPUT_VALUE_INVALID, fieldErrors);
    }

    @ExceptionHandler(UserAlreadyExistsException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected ErrorResponse handleUserAlreadyExistsException(UserAlreadyExistsException e) {
        ErrorCode errorCode = ErrorCode.USER_ALREADY_EXISTS;
        log.error("User Id Duplication : " + e.getUserId());
        return buildError(errorCode);
    }

    @ExceptionHandler(UserNotFoundException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected ErrorResponse handleUserNotFoundException(UserNotFoundException e) {
        ErrorCode errorCode = ErrorCode.USER_NOT_FOUND;
        return buildError(errorCode);
    }

    @ExceptionHandler(UserPasswordNotEqualsException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected ErrorResponse handleUserPasswordNotEqualsException(UserPasswordNotEqualsException e) {
        ErrorCode errorCode = ErrorCode.USER_PASSWORD_FAIL;
        return buildError(errorCode);
    }

    private List<ErrorResponse.FieldError> getFieldErrors(BindingResult bindingResult) {
        List<FieldError> fieldErrors = bindingResult.getFieldErrors();
        return fieldErrors.parallelStream()
                .map(err -> new ErrorResponse.FieldError(err.getField(), (String) err.getRejectedValue(), err.getDefaultMessage()
                ))
                .collect(Collectors.toList());
    }

    private ErrorResponse buildFieldErrors(ErrorCode errorCode, List<ErrorResponse.FieldError> fieldErrors) {
        return new ErrorResponse(errorCode.getMessage(), errorCode.getCode(), errorCode.getStatus(), fieldErrors);
    }

    private ErrorResponse buildError(ErrorCode errorCode) {
        return new ErrorResponse(errorCode.getMessage(), errorCode.getCode(), errorCode.getStatus(), List.of());
    }
}

File: migration-to-kotlin/src/main/kotlin/com/example/practice/error/ExceptionHandlingController.kt
package com.example.practice.error

import com.example.practice.exception.UserAlreadyExistsException
import com.example.practice.exception.UserNotFoundException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.validation.BindingResult
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ExceptionHandlingController {
    private val log : Logger = LoggerFactory.getLogger(ExceptionHandlingController::class.java)

    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun handleMethodArgumentNotValidException(e: MethodArgumentNotValidException) : ErrorResponse {
        log.error("Input value invalidation")
        val fieldErrors : List<ErrorResponse.FieldError> = getFieldErrors(e.bindingResult)
        return buildFieldErrors(ErrorCode.INPUT_VALUE_INVALID, fieldErrors)
    }

    @ExceptionHandler(UserAlreadyExistsException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun handleUserAlreadyExistsException(e: UserAlreadyExistsException) : ErrorResponse {
        val errorCode = ErrorCode.UID_DUPLICATION
        log.error("User ID Duplication : ${e.userId}")
        return buildError(errorCode)
    }

    @ExceptionHandler(UserNotFoundException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun handleUserNotFoundException(e: UserNotFoundException) : ErrorResponse {
        log.error("User Not Found : ${e.userId}")
        return buildError(ErrorCode.USER_NOT_FOUND)
    }

    private fun getFieldErrors(bindingResult: BindingResult) : List<ErrorResponse.FieldError> {
        val fieldErrors : List<FieldError> = bindingResult.fieldErrors
        return fieldErrors.parallelStream()
            .map { ErrorResponse.FieldError(
                field = it.field,
                value = it.rejectedValue as String,
                reason = it.defaultMessage!!
            ) }
            .toList()
    }

    private fun buildFieldErrors(errorCode: ErrorCode, fieldErrors: List<ErrorResponse.FieldError>) : ErrorResponse {
        return ErrorResponse(
            code = errorCode.code,
            status = errorCode.status,
            message = errorCode.message,
            errors = fieldErrors
        )
    }

    private fun buildError(errorCode: ErrorCode) : ErrorResponse {
        return ErrorResponse(
            code = errorCode.code,
            status = errorCode.status,
            message = errorCode.message
        )
    }
}